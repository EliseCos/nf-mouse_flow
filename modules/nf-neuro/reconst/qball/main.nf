process RECONST_QBALL {
    tag "$meta.id"
    label 'process_medium'

    container "scilus/scilpy:2.2.0_cpu"

    input:
    tuple val(meta), path(dwi), path(bval), path(bvec), path(mask)

    output:
    tuple val(meta), path("*__qball.nii.gz")        , emit: qball, optional: true
    tuple val(meta), path("*__gfa.nii.gz")          , emit: gfa, optional: true
    tuple val(meta), path("*__peaks.nii.gz")        , emit: peaks, optional: true
    tuple val(meta), path("*__peak_indices.nii.gz") , emit: peak_indices, optional: true
    tuple val(meta), path("*__nufo.nii.gz")         , emit: nufo, optional: true
    tuple val(meta), path("*__a_power.nii.gz")      , emit: a_power, optional: true
    path "versions.yml"                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def b0_threshold = task.ext.b0_threshold ? " --b0_threshold " + task.ext.b0_threshold : ""
    def sh_order =  task.ext.sh_order ? " --sh_order " + task.ext.sh_order : ""
    def processes = task.cpu ? " --processes " + task.cpu : "--processes 1"

    if ( mask ) args += " --mask $mask"
    if ( task.ext.gfa ) args += " --gfa ${prefix}__gfa.nii.gz"
    if ( task.ext.peaks ) args += " --peaks ${prefix}__peaks.nii.gz"
    if ( task.ext.peak_indices ) args += " --peak_indices ${prefix}__peak_indices.nii.gz"
    if ( task.ext.qball) args += " --sh ${prefix}__qball.nii.gz"
    if ( task.ext.nufo) args += " --nufo ${prefix}__nufo.nii.gz"
    if ( task.ext.a_power) args += " --a_power ${prefix}__a_power.nii.gz"
    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    scil_qball_metrics $dwi $bval $bvec --not_all $args $b0_threshold $sh_order $processes

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    scil_qball_metrics -h

    touch ${prefix}__gfa.nii.gz
    touch ${prefix}__peaks.nii.gz
    touch ${prefix}__peak_indices.nii.gz
    touch ${prefix}__sh.nii.gz
    touch ${prefix}__nufo.nii.gz
    touch ${prefix}__a_power.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
