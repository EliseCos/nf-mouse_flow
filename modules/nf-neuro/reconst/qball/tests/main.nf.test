nextflow_process {

    name "Test Process RECONST_QBALL"
    script "../main.nf"
    process "RECONST_QBALL"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfneuro"
    tag "reconst"
    tag "reconst/qball"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "DWIss1000-dir32.zip", "segmentation.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("reconst - qball_csa") {
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        dwi: it.simpleName == "DWIss1000-dir32"
                        segmentation: it.simpleName == "segmentation"
                    }
                ch_dwi = ch_split_test_data.dwi.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]
                }
                ch_mask = ch_split_test_data.segmentation.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/brainmask/brainmask.nii.gz")
                    ]
                }
                input[0] = ch_dwi
                    .join(ch_mask)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    niftiMD5SUM(process.out.gfa.get(0).get(1), 6),
                    file(process.out.peaks.get(0).get(1)).name,
                    file(process.out.peak_indices.get(0).get(1)).name,
                    niftiMD5SUM(process.out.sh.get(0).get(1), 6),
                    niftiMD5SUM(process.out.nufo.get(0).get(1), 6),
                    niftiMD5SUM(process.out.a_power.get(0).get(1), 6),
                    process.out.versions
                ).match() }
            )
        }
    }

    test("reconst - qball_gfa") {
        config "./nextflow_light.config"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        dwi: it.simpleName == "DWIss1000-dir32"
                        segmentation: it.simpleName == "segmentation"
                    }
                input[0] = ch_split_test_data.dwi.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec"),
                        []
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    niftiMD5SUM(process.out.gfa.get(0).get(1), 6),
                    process.out.versions
                ).match() }
            )
        }
    }

    test("reconst - stub-run qball") {
        tag "stub"
        options "-stub-run"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        dwi: it.simpleName == "DWIss1000-dir32"
                        segmentation: it.simpleName == "segmentation"
                    }
                input[0] = ch_split_test_data.dwi.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec"),
                        []
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }
}
