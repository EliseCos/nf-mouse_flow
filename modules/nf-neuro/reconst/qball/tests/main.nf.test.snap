{
    "reconst - qball_csa": {
        "content": [
            "test__gfa.nii.gz:md5:header,c9e720b9a8acfe45a617e6486ebb8de0,data,776c03e1654897b87c29b9f940a255ca",
            "test__peaks.nii.gz",
            "test__peak_indices.nii.gz",
            "test__sh.nii.gz:md5:header,76a9953111f97786bd14a451c61529e7,data,bdd9a72f76558900d2750f3e88116bab",
            "test__nufo.nii.gz:md5:header,ccd78a751ac6eddb4cc7e037e1c82a41,data,0c6a62153a737d93bbe86d86b418adbb",
            "test__a_power.nii.gz:md5:header,c9e720b9a8acfe45a617e6486ebb8de0,data,d6b4b3f81f9b7f15e6f8de4a4441d563",
            [
                "versions.yml:md5,467924bf83804809033ee79becb0af51"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T03:40:39.803464351"
    },
    "reconst - qball_gfa": {
        "content": [
            "test__gfa.nii.gz:md5:header,c9e720b9a8acfe45a617e6486ebb8de0,data,776c03e1654897b87c29b9f940a255ca",
            [
                "versions.yml:md5,467924bf83804809033ee79becb0af51"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T02:58:44.76992447"
    },
    "reconst - stub-run qball": {
        "content": [
            [
                "versions.yml:md5,467924bf83804809033ee79becb0af51"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T02:24:26.378026962"
    }
}